source/common/stats/symbol_table_impl.cc

```diff
@@ -117,86 +117,144 @@ std::pair<uint64_t, size_t> SymbolTableImpl::Encoding::decodeNumber(const uint8_
-      array += symbol_consumed.second;
+      array_ += length_consumed.second;
+      size_ -= length_consumed.second;
+      ASSERT(size_ >= length);
+      string_view_ = absl::string_view(reinterpret_cast<const char*>(array_), length);
+      size_ -= length;
+      array_ += length;
+#ifndef NDEBUG
+      last_type_ = TokenType::StringView;
+#endif
+      return TokenType::StringView;
     }
+    std::pair<uint64_t, size_t> symbol_consumed = decodeNumber(array_);
+    symbol_ = symbol_consumed.first;
+    size_ -= symbol_consumed.second;
+    array_ += symbol_consumed.second;
+#ifndef NDEBUG
+    last_type_ = TokenType::Symbol;
+#endif
+    return TokenType::Symbol;
   }
-}
 
-bool StatName::startsWith(StatName symbolic_prefix) const {
-  bool ret = true;
-  std::vector<Symbol> prefix_symbols;
+  /** @return the current string_view -- only valid to call if next()==TokenType::StringView */
+  absl::string_view stringView() const {
+#ifndef NDEBUG
+    ASSERT(last_type_ == TokenType::StringView);
+#endif
+    return string_view_;
+  }
 
-  // Decode the prefix as a StatNameVec.
-  SymbolTableImpl::Encoding::decodeTokens(
-      symbolic_prefix.data(), symbolic_prefix.dataSize(),
-      [&prefix_symbols](Symbol symbol) { prefix_symbols.push_back(symbol); },
-      [&ret](absl::string_view) { ret = false; });
+  /** @return the current symbol -- only valid to call if next()==TokenType::Symbol */
+  Symbol symbol() const {
+#ifndef NDEBUG
+    ASSERT(last_type_ == TokenType::Symbol);
+#endif
+    return symbol_;
+  }
+
+private:
+  const uint8_t* array_;
+  size_t size_;
+  absl::string_view string_view_;
+  Symbol symbol_;
+#ifndef NDEBUG
+  TokenType last_type_{TokenType::End};
+#endif
+};
+
+void SymbolTableImpl::Encoding::decodeTokens(
+    StatName stat_name, const std::function<void(Symbol)>& symbol_token_fn,
+    const std::function<void(absl::string_view)>& string_view_token_fn) {
+  TokenIter iter(stat_name);
+  TokenIter::TokenType type;
+  while ((type = iter.next()) != TokenIter::TokenType::End) {
+    if (type == TokenIter::TokenType::StringView) {
+      string_view_token_fn(iter.stringView());
+    } else {
+      ASSERT(type == TokenIter::TokenType::Symbol);
+      symbol_token_fn(iter.symbol());
+    }
+  }
+}
 
-  // If there any dynamic components, our string_view lambda will be called, and
-  // then we'll return false for simplicity. We don't have a current need for
-  // prefixes to be expressed dynamically, and handling that case would add
-  // complexity.
-  if (!ret) {
-    return false;
+bool StatName::startsWith(StatName prefix) const {
+  using TokenIter = SymbolTableImpl::Encoding::TokenIter;
+  TokenIter prefix_iter(prefix);
+  TokenIter this_iter(*this);
+  while (true) {
+    TokenIter::TokenType prefix_type = prefix_iter.next();
+    TokenIter::TokenType this_type = this_iter.next();
+    if (prefix_type == TokenIter::TokenType::End) {
+      break; // "a.b.c" starts with "a.b" or "a.b.c"
```

Comment:
Maybe just use return here? It's a bit confusing having both break and return control flows here imo